proposed/activity-improvements.md

```diff
@@ -0,0 +1,290 @@
+
+        /// Consumer activity represents output received from an external component.
+        Consumer = 5,
+    }
+}
+```
+
+ ### ActivityContext
+
+```C#
+    /// ActivityContext representation conforms to the w3c TraceContext specification. It contains two identifiers
+    /// a TraceId and a SpanId - along with a set of common TraceFlags and system-specific TraceState values.
+    public readonly struct ActivityContext : IEquatable<ActivityContext>
+    {
+        public ActivityContext(ActivityTraceId traceId, ActivitySpanId spanId, ActivityTraceFlags traceFlags, string? traceState = null)
+        public ActivityTraceId TraceId { get; }
+        public ActivitySpanId SpanId { get; }
+        public ActivityTraceFlags TraceFlags { get; }
+        public string? TraceState { get; }
+
+        public static bool operator ==(ActivityContext context1, ActivityContext context2)
+        public static bool operator !=(ActivityContext context1, ActivityContext context2)
+        public bool Equals(ActivityContext context)
+        public override bool Equals(object? obj)
+        public override int GetHashCode()
+    }
+```
+
+ ### ActivityLink
+```C#
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// Activity may be linked to zero or more other <see cref="ActivityContext"/> that are causally related.
+    /// Links can point to ActivityContexts inside a single Trace or across different Traces.
+    /// Links can be used to represent batched operations where a Activity was initiated by multiple initiating Activities,
+    /// each representing a single incoming item being processed in the batch.
+    public readonly struct ActivityLink
+    {
+        public ActivityLink(ActivityContext context) 
+        public ActivityLink(ActivityContext context, IDictionary<string, object>? attributes)
+        public ActivityContext Context { get; }
+        public IDictionary<string, object>? Attributes { get; }
+    }
+}
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable 
+    {
+        ....
+        public System.Diagnostics.ActivityKind Kind { get; set; }
+        public System.Collections.Generic.IEnumerable<ActivityLink> Links { get; }
+        public System.Diagnostics.Activity AddLink(ActivityLink link) 
+
+        public static System.Diagnostics.Activity? StartNew(string name) 
+        public static System.Diagnostics.Activity? StartNew(string name, ActivityContext context, IEnumerable<ActivityLink>? links = null, DateTimeOffset statTime = default) 
+        public static System.Diagnostics.Activity? StartNew(string name, ActivityContext context, ActivityKind kind, IEnumerable<System.Collections.Generic.KeyValuePair<string,string>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default) 
+
+        public bool TryGetContext(out ActivityContext context)
+
+        public static IDisposable StartListening(
```

Comment:
Moving thread from [old issue](https://github.com/dotnet/runtime/issues/31373#issuecomment-600908953)

@SergeyKanzhelev wrote:

> Having specific callback for sampling instead of having the sampling decision made in creation callback is not aligned with OpenTelemetry spec. Some features like attributes returned by Sampler to be associated with the Span will simply not work.

Reply:
I think we are open to alternate forms of the callback including changing its name and signature. For example it could return an enum?

In terms of the attribute how common of case is it and what are they used for? One option is that the attributes could be applied during onActivityStarted. Other possibilities are that the attribute collection passed to the call is mutable or there is a second collection that is returned/passed via out param.